### PR TITLE
Interactive execution

### PR DESCRIPTION
**Pull request by leynos**
## Summary by Sourcery

Implement interactive container exec support with attached/detached modes, TTY handling, and exit-code propagation through the CLI.

New Features:
- Add engine exec module with attached/detached execution modes, TTY control, and terminal resize handling.
- Wire `podbot exec` to Docker via EngineConnector, supporting a new `--detach` flag and propagating container command exit codes.
- Introduce BDD feature scenarios and helpers for interactive execution behaviour across attached/detached modes.

Enhancements:
- Refine CLI command dispatch to return structured command outcomes and normalize process exit codes.
- Extend engine connection public API with exec-related client traits and types for better testability and abstraction.
- Document interactive exec semantics, usage patterns, and roadmap completion in design and user guides.

Documentation:
- Expand user guide and design docs to cover interactive exec semantics, terminal behaviour, and exit-code propagation, and record the execution plan in a new execplan document.
- Update the project roadmap to mark interactive execution tasks as complete.

Tests:
- Add unit tests for the exec lifecycle, including error mapping, TTY resize handling, and exit-code capture.
- Add BDD-style tests and feature files for interactive exec behaviour, including attached, detached, and failure scenarios.

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[src/engine/connection/exec/tests.rs](https://github.com/leynos/podbot/pull/41/files/81a756d2ae76a97c31ca01fdcf2ce36874334e11#diff-e7129f498e162ff3e0b1df8e858578e1d7abc20e6c74268be8a3c618d0df2f70)

Comment on file

~~~rust
fn exec_async_attached_skips_resize_when_tty_disabled(runtime: tokio::runtime::Runtime) {
    let mut client = MockExecClient::new();
    client
        .expect_create_exec()
        .times(1)
        .returning(|_, options| {
            assert_eq!(options.tty, Some(false));
            Box::pin(async {
                Ok(CreateExecResults {
                    id: String::from("exec-5"),
                })
            })
        });
    client.expect_start_exec().times(1).returning(|_, options| {
        assert_eq!(
            options,
            Some(StartExecOptions {
                detach: false,
                tty: false,
                output_capacity: None
            })
        );
        let output_stream = stream::iter(Vec::<Result<LogOutput, BollardError>>::new());
        Box::pin(async move {
            Ok(bollard::exec::StartExecResults::Attached {
                output: Box::pin(output_stream),
                input: Box::pin(tokio::io::sink()),
            })
        })
    });
    client.expect_resize_exec().never();
    client.expect_inspect_exec().times(1).returning(|_| {
        Box::pin(async {
            Ok(bollard::models::ExecInspectResponse {
                running: Some(false),
                exit_code: Some(0),
                ..bollard::models::ExecInspectResponse::default()
            })
        })
    });

    let request = ExecRequest::new(
        "sandbox-123",
        vec![String::from("echo"), String::from("hello")],
        ExecMode::Attached,
    )
    .expect("attached request should build")
    .with_tty(false);
    let terminal_size_provider = StubTerminalSizeProvider {
        terminal_size: Some(TerminalSize {
            width: 80,
            height: 24,
        }),
    };

    let result = runtime.block_on(EngineConnector::exec_async_with_terminal_size_provider(
        &client,
        &request,
        &terminal_size_provider,
    ));
    assert!(
        result.is_ok(),
        "attached execution should succeed: {result:?}"
    );
}
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/75835/delta?repo-id=1423864&review-id=41&biomarker=Code+Duplication&filename=src%2Fengine%2Fconnection%2Fexec%2Ftests.rs&method=)
The module contains 5 functions with similar structure: exec_async_attached_calls_resize_when_tty_enabled,exec_async_attached_propagates_resize_failures,exec_async_attached_skips_resize_when_tty_disabled,exec_request_rejects_blank_container_id and 1 more functions

File: src/engine/connection/exec/attached.rs [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+
+use bollard::container::LogOutput;
+use bollard::errors::Error as BollardError;
+use futures_util::{Stream, StreamExt};
+use tokio::io::{AsyncReadExt, AsyncWrite, AsyncWriteExt};
+use tokio::time::sleep;
+
+use super::terminal::{
+    TerminalSizeProvider, maybe_sigwinch_listener, resize_exec_to_current_terminal_async,
+    wait_for_sigwinch,
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/engine/connection/exec/attached.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

~~~
Please address the comments from this code review:

## Overall Comments
- In `exec_in_container` you create a dedicated Tokio runtime instead of using the `EngineConnector::exec` helper that already accepts a `Handle`; consider centralising runtime management (e.g. a single runtime in `main` passed down) so exec paths share the same executor and you avoid per-command runtime creation.
- `normalize_process_exit_code` currently does `i32::try_from(code).unwrap_or(1)`, which silently maps any out-of-range exit code to `1`; it may be clearer to clamp or explicitly document/handle negative and oversized exit codes so callers understand how container exit codes are translated into process exits.
- The terminal sizing logic relies on spawning the external `stty` command and silently ignores failures by returning `None`; if this is expected only on specific platforms or environments, consider tightening the conditions under which `stty` is used or surfacing a clearer error/log path so unexpected absence of `stty` doesn’t make interactive behaviour harder to diagnose.

## Individual Comments

### Comment 1
<location> `src/engine/connection/exec/attached.rs:40-41` </location>
<code_context>
+    input: Pin<Box<dyn AsyncWrite + Send>>,
+    size_provider: &P,
+) -> Result<(), PodbotError> {
+    let stdin_task = tokio::spawn(async move { forward_stdin_to_exec_async(input).await });
+    let mut stdout = tokio::io::stdout();
+    let mut stderr = tokio::io::stderr();
+
</code_context>

<issue_to_address>
**issue (bug_risk):** The spawned stdin forwarding task can be leaked if the function returns early before the cleanup block.

If `maybe_sigwinch_listener(request)?` or the initial `resize_exec_to_current_terminal_async(...)` fail, the function returns before the cleanup that aborts/joins `stdin_task`, leaving a background task still reading stdin and writing to a dead exec input stream.

Consider ensuring `stdin_task` is always cancelled/joined on early returns by either:
- Spawning it only after all fallible setup completes, or
- Using a small RAII/`Drop` helper to abort/join it on scope exit, or
- Wrapping the body in a `result = (|| async { ... })().await` pattern and doing the cleanup after matching on `result`.

This avoids leaking a task that continues consuming stdin after the exec session fails to start.
</issue_to_address>

### Comment 2
<location> `src/engine/connection/exec/tests.rs:45-46` </location>
<code_context>
+    tokio::runtime::Runtime::new().expect("runtime creation should succeed")
+}
+
+#[rstest]
+fn exec_request_rejects_empty_command() {
+    let result = ExecRequest::new("sandbox", vec![], ExecMode::Attached);
+    assert!(
</code_context>

<issue_to_address>
**suggestion (testing):** Add a test to cover rejection of commands that contain blank/whitespace-only entries

`validate_command` also rejects commands where at least one entry is non-empty but only whitespace. Currently only the empty-command path is covered. Please add a test (e.g. `exec_request_rejects_blank_command_entries`) that calls `ExecRequest::new("sandbox", vec!["   ".into(), "echo".into()], ExecMode::Attached)` and asserts it returns `ConfigError::InvalidValue { field: "command", .. }` to lock in this behaviour.
</issue_to_address>

### Comment 3
<location> `src/engine/connection/exec/tests.rs:74-83` </location>
<code_context>
+        return Ok(());
+    };
+
+    client
+        .resize_exec(
+            exec_id,
</code_context>

<issue_to_address>
**suggestion (testing):** Consider a negative test where `resize_exec` fails to ensure error propagation is covered

Current tests cover the happy path and the case where resize is skipped when TTY is disabled, but none exercise the failure path of `resize_exec`. Please add a test that forces `resize_exec` to return an error and asserts that `exec_async_with_terminal_size_provider` returns `ContainerError::ExecFailed` to validate error propagation for terminal resize failures.
</issue_to_address>

### Comment 4
<location> `docs/users-guide.md:104` </location>
<code_context>
+  targets, podbot also listens for `SIGWINCH` and propagates window-size
+  changes. Detached mode, or attached mode with TTY disabled, does not register
+  a resize listener.
+- Podbot polls exec status until the command exits, then uses the daemon
+  exit code as the CLI outcome: `0` returns success and non-zero codes are
+  propagated as the process exit status.
</code_context>

<issue_to_address>
**nitpick (typo):** Use consistent capitalisation for "podbot" in this bullet point.

In this section the tool name is written as `podbot` elsewhere; please update `Podbot polls exec status` to `podbot polls exec status` for consistency.

```suggestion
- podbot polls exec status until the command exits, then uses the daemon
```
</issue_to_address>

### Comment 5
<location> `src/engine/connection/exec/attached.rs:175` </location>
<code_context>
+    Ok(true)
+}
+
+async fn forward_stdin_to_exec_async(mut input: Pin<Box<dyn AsyncWrite + Send>>) -> io::Result<()> {
+    let mut stdin = tokio::io::stdin();
+    let mut buffer = [0_u8; 8192];
</code_context>

<issue_to_address>
**issue (complexity):** Consider simplifying the attached exec stdin forwarding by allowing `Unpin` on the writer so you can use `AsyncWriteExt`/`tokio::io::copy` and by relaxing the stdin task join handling to avoid custom poll loops and error plumbing.

You can drop a fair bit of low‑level async I/O plumbing without changing behaviour by leaning on `AsyncWriteExt`/`tokio::io::copy` and slightly adjusting types.

### 1. Remove `write_all_pinned_async` / `flush_pinned_async` / manual poll loop

The only reason you need manual `poll_fn` is that `input` is `Pin<Box<dyn AsyncWrite + Send>>`. If you allow `Unpin` on the trait object, you can use `write_all`/`flush` directly and/or `tokio::io::copy`.

Change the `run_attached_session_async` signature:

```rust
pub(super) async fn run_attached_session_async<
    C: ContainerExecClient,
    P: TerminalSizeProvider,
>(
    client: &C,
    request: &ExecRequest,
    exec_id: &str,
    mut output: Pin<Box<dyn Stream<Item = Result<LogOutput, BollardError>> + Send>>,
    input: Box<dyn AsyncWrite + Send + Unpin>, // <- add Unpin, drop Pin
    size_provider: &P,
) -> Result<(), PodbotError> {
    let stdin_task = tokio::spawn(async move { forward_stdin_to_exec_async(input).await });
    // ...
}
```

Then simplify `forward_stdin_to_exec_async` to use `tokio::io::copy` or at least `write_all`/`flush`:

```rust
async fn forward_stdin_to_exec_async(
    mut input: Box<dyn AsyncWrite + Send + Unpin>,
) -> io::Result<()> {
    let mut stdin = tokio::io::stdin();
    tokio::io::copy(&mut stdin, &mut input).await?;
    input.flush().await
}
```

If you want to keep the explicit buffering semantics instead of `copy`:

```rust
async fn forward_stdin_to_exec_async(
    mut input: Box<dyn AsyncWrite + Send + Unpin>,
) -> io::Result<()> {
    let mut stdin = tokio::io::stdin();
    let mut buffer = [0_u8; 8192];

    loop {
        let bytes_read = stdin.read(&mut buffer).await?;
        if bytes_read == 0 {
            break;
        }
        input.write_all(&buffer[..bytes_read]).await?;
    }

    input.flush().await
}
```

With this, `write_all_pinned_async` and `flush_pinned_async` can be deleted entirely.

If `Unpin` on the trait object is not acceptable, you can wrap it in a newtype that implements `AsyncWrite` + `Unpin` by delegating to the pinned inner value; that keeps the call sites using `write_all` instead of custom poll loops.

### 2. Simplify stdin task abort/join handling

If you don’t need to distinguish between “cancelled” vs “error” when joining the stdin task, the explicit `JoinError` inspection can be dropped:

```rust
stdin_task.abort();
let _ = stdin_task.await; // ignore cancellation/error on shutdown
```

If you do care about logging errors but not about propagating them, you can move that into the spawned task:

```rust
let stdin_task = tokio::spawn(async move {
    if let Err(error) = forward_stdin_to_exec_async(input).await {
        // log at debug/info; no need to propagate to caller
        tracing::debug!(%error, "stdin forwarding task ended with error");
    }
});

// ... later:
stdin_task.abort();
let _ = stdin_task.await;
```

This keeps the top‑level control flow focused on the exec session lifecycle instead of join semantics, while preserving the current behaviour.
</issue_to_address>

### Comment 6
<location> `src/engine/connection/exec/attached.rs:32` </location>
<code_context>
+    clippy::integer_division_remainder_used,
+    reason = "false positive triggered inside tokio::select! expansion"
+)]
+pub(super) async fn run_attached_session_async<C: ContainerExecClient, P: TerminalSizeProvider>(
+    client: &C,
+    request: &ExecRequest,
</code_context>

<issue_to_address>
**issue (review_instructions):** `run_attached_session_async` is quite large and contains multiple nested control flows, which likely pushes its cyclomatic complexity beyond the requested maximum of 12.

This function coordinates stdin forwarding, output streaming, SIGWINCH handling, and initial resize logic all in one place, with a loop, cfg-gated `tokio::select!`, and several error paths. That combination suggests its cyclomatic complexity is above the target of 12.

Consider extracting some of the responsibilities into smaller helpers (e.g. a function to handle a single output tick, another to manage SIGWINCH/resize behaviour, or to encapsulate the stdin task lifecycle). That should reduce the complexity of this top-level orchestration function while keeping the behaviour intact.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.rs`

**Instructions:**
Seek to keep the cyclomatic complexity of functions no more than 12.

</details>
</issue_to_address>

### Comment 7
<location> `docs/execplans/2-4-1-interactive-execution.md:70` </location>
<code_context>
+  Severity: medium. Likelihood: medium. Mitigation: isolate size capture behind
+  a small helper seam and test with injected values.
+- Risk: attached stream forwarding can create deadlocks if stdin/stdout tasks do
+  not close correctly. Severity: high. Likelihood: medium. Mitigation: model IO
+  forwarding as explicit tasks with clear shutdown ordering and unit tests for
+  close/error paths.
</code_context>

<issue_to_address>
**issue (review_instructions):** The acronym "IO" is used without being expanded on first use.

Please expand "IO" the first time it appears, for example "input/output (IO) forwarding", and then use the abbreviated form afterwards.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.md`

**Instructions:**
Define uncommon acronyms on first use.

</details>
</issue_to_address>
~~~
